File: macosx/DragOverlayView.m
/******************************************************************************
 * $Id$
 *
 * Copyright (c) 2007-2008 Transmission authors and contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *****************************************************************************/

#import "DragOverlayView.h"
#import "NSBezierPathAdditions.h"

#define PADDING 10.0
#define ICON_WIDTH 64.0

@implementation DragOverlayView

- (id) initWithFrame: (NSRect) frame
{
    if ((self = [super initWithFrame: frame]))
    {
        //create badge
        NSRect badgeRect = NSMakeRect(0, 0, 325.0, 84.0);
        NSBezierPath * bp = [NSBezierPath bezierPathWithRoundedRect: badgeRect radius: 15.0];
        
        fBackBadge = [[NSImage alloc] initWithSize: badgeRect.size];
        [fBackBadge lockFocus];
        
        [[NSColor colorWithCalibratedWhite: 0.0 alpha: 0.75] set];
        [bp fill];
        
        [fBackBadge unlockFocus];
        
        //create attributes
        NSShadow * stringShadow = [[NSShadow alloc] init];
        [stringShadow setShadowOffset: NSMakeSize(2.0, -2.0)];
        [stringShadow setShadowBlurRadius: 4.0];
        
        NSFont * bigFont = [[NSFontManager sharedFontManager] convertFont:
                                [NSFont fontWithName: @"Lucida Grande" size: 18.0] toHaveTrait: NSBoldFontMask],
                * smallFont = [NSFont fontWithName: @"Lucida Grande" size: 14.0];
        
        NSMutableParagraphStyle * paragraphStyle = [[NSParagraphStyle defaultParagraphStyle] mutableCopy];
        [paragraphStyle setLineBreakMode: NSLineBreakByTruncatingTail];
        
        fMainLineAttributes = [[NSDictionary alloc] initWithObjectsAndKeys:
                                [NSColor whiteColor], NSForegroundColorAttributeName,
                                bigFont, NSFontAttributeName, stringShadow, NSShadowAttributeName,
                                paragraphStyle, NSParagraphStyleAttributeName, nil];
        
        fSubLineAttributes = [[NSDictionary alloc] initWithObjectsAndKeys:
                                [NSColor whiteColor], NSForegroundColorAttributeName,
                                smallFont, NSFontAttributeName, stringShadow, NSShadowAttributeName,
                                paragraphStyle, NSParagraphStyleAttributeName, nil];
        
        [stringShadow release];
        [paragraphStyle release];
    }
    return self;
}

- (void) dealloc
{
    [fBackBadge release];
    [fBadge release];
    
    [fMainLineAttributes release];
    [fSubLineAttributes release];
    
    [super dealloc];
}

- (void) setOverlay: (NSImage *) icon mainLine: (NSString *) mainLine subLine: (NSString *) subLine
{
    [fBadge release];
    fBadge = [fBackBadge copy];
    NSSize badgeSize = [fBadge size];
    
    [fBadge lockFocus];
    
    //place icon
    [icon drawInRect: NSMakeRect(PADDING, (badgeSize.height - ICON_WIDTH) * 0.5, ICON_WIDTH, ICON_WIDTH) fromRect: NSZeroRect
            operation: NSCompositeSourceOver fraction: 1.0];
    
    //place main text
    NSSize mainLineSize = [mainLine sizeWithAttributes: fMainLineAttributes];
    NSSize subLineSize = [subLine sizeWithAttributes: fSubLineAttributes];
    
    NSRect lineRect = NSMakeRect(PADDING + ICON_WIDTH + 5.0,
                        (badgeSize.height + (subLineSize.height + 2.0 - mainLineSize.height)) * 0.5,
                        badgeSize.width - (PADDING + ICON_WIDTH + 2.0) - PADDING, mainLineSize.height);
    [mainLine drawInRect: lineRect withAttributes: fMainLineAttributes];
    
    //place sub text
    lineRect.origin.y -= subLineSize.height + 2.0;
    lineRect.size.height = subLineSize.height;
    [subLine drawInRect: lineRect withAttributes: fSubLineAttributes];
    
    [fBadge unlockFocus];
    
    [self setNeedsDisplay: YES];
}

-(void) drawRect: (NSRect) rect
{
    if (fBadge)
    {
        NSRect frame = [self frame];
        NSSize imageSize = [fBadge size];
        [fBadge compositeToPoint: NSMakePoint((frame.size.width - imageSize.width) * 0.5,
                    (frame.size.height - imageSize.height) * 0.5) operation: NSCompositeSourceOver];
    }
}

@end

File: macosx/DragOverlayView.swift
// This file Copyright © 2007-2023 Transmission authors and contributors.
// It may be used under the MIT (SPDX: MIT) license.
// License text can be found in the licenses/ folder.

import AppKit

class DragOverlayView: NSView {
    private static let kPadding: CGFloat = 10.0
    private static let kIconWidth: CGFloat = 64.0

    private var fBadge: NSImage?

    private var fMainLineAttributes: [NSAttributedString.Key: Any]
    private var fSubLineAttributes: [NSAttributedString.Key: Any]

    override init(frame frameRect: NSRect) {
        // create attributes
        let stringShadow = NSShadow()
        stringShadow.shadowOffset = NSSize(width: 2.0, height: -2.0)
        stringShadow.shadowBlurRadius = 4.0

        let bigFont = NSFont.boldSystemFont(ofSize: 18.0)
        let smallFont = NSFont.systemFont(ofSize: 14.0)

        // swiftlint:disable:next force_cast
        let paragraphStyle = NSParagraphStyle.default.mutableCopy() as! NSMutableParagraphStyle
        paragraphStyle.lineBreakMode = .byTruncatingMiddle

        fMainLineAttributes = [
            .foregroundColor: NSColor.white,
            .font: bigFont,
            .shadow: stringShadow,
            .paragraphStyle: paragraphStyle,
        ]
        fSubLineAttributes = [
            .foregroundColor: NSColor.white,
            .font: smallFont,
            .shadow: stringShadow,
            .paragraphStyle: paragraphStyle,
        ]
        super.init(frame: frameRect)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    @objc
    func setOverlay(_ icon: NSImage, mainLine: String, subLine: String) {
        // create badge
        let badgeRect = NSRect(x: 0.0, y: 0.0, width: 325.0, height: 84.0)

        fBadge = NSImage(size: badgeRect.size)
        fBadge?.lockFocus()

        let bp = NSBezierPath(roundedRect: badgeRect, xRadius: 15.0, yRadius: 15.0)
        NSColor(calibratedWhite: 0.0, alpha: 0.75).set()
        bp.fill()

        // place icon
        icon.draw(in: NSRect(x: Self.kPadding, y: (badgeRect.height - Self.kIconWidth) * 0.5, width: Self.kIconWidth, height: Self.kIconWidth),
                  from: .zero,
                  operation: .sourceOver,
                  fraction: 1.0)

        // place main text
        let mainLineSize = mainLine.size(withAttributes: fMainLineAttributes)
        let subLineSize = subLine.size(withAttributes: fSubLineAttributes)

        var lineRect = NSRect(x: Self.kPadding + Self.kIconWidth + 5.0,
                              y: (badgeRect.height + (subLineSize.height + 2.0 - mainLineSize.height)) * 0.5,
                              width: badgeRect.width - (Self.kPadding + Self.kIconWidth + 2.0) - Self.kPadding,
                              height: mainLineSize.height)
        mainLine.draw(in: lineRect, withAttributes: fMainLineAttributes)

        // place sub text
        lineRect.origin.y -= subLineSize.height + 2.0
        lineRect.size.height = subLineSize.height
        subLine.draw(in: lineRect, withAttributes: fSubLineAttributes)

        fBadge?.unlockFocus()

        needsDisplay = true
    }

    override func draw(_ dirtyRect: NSRect) {
        if let fBadge = fBadge {
            let frame = self.frame
            let imageSize = fBadge.size
            fBadge.draw(at: NSPoint(x: (frame.width - imageSize.width) * 0.5, y: (frame.height - imageSize.height) * 0.5),
                        from: .zero,
                        operation: .sourceOver,
                        fraction: 1.0)
        }
    }
}
